### PR TITLE
Fixes issue #1557 Changes to support switching to bigquery. 

### PR DESCRIPTION
**Pull request by jonespm**
The queries currently aren't working but it feels like it might be making progress!

- [x] Update and test query `metadata`
- [x] Update and test query `user`
- [x] Update and test query `assignment_groups`
- [x] Update and test query `assignment`
- [x] Update and test query `assignment_weight`
- [x] Update and test query `term`
- [x] Update and test query `course`
- [x] Update and test query `resource`
- [x] Update and test query `submission`
- [x] Update and test query 'submission_with_avg_score' (Combined with submission)
- [x] Clean up postgres package, DATA_WAREHOUSE setting and code using it from db_util if. this is no longer necessary. 
- [x] See if we need to do anything about `timestamp without time zone` to keep dates correct
- [x] Compare old and new databases to ensure accuracy of data
- [x] Double check that start and end date are being correctly populated into courses
- [x] Remove utility function and just bind BQ parameters directly
- [x] Remove left over DATA_WAREHOUSE references

### CODE REVIEW

**Comment by pushyamig** [comment]
The queries are Postgres Syntax so needs some tweaking to BQ format, so that is the reason might now be working. I also feel the submission query might perform bit better with BQ may be not need to create a Temp table(Needs to play around with it)

**Comment by jonespm** [comment]
I put this up as draft right now because @zqian wanted to see it. No need to review any code here as it's really just a work in progress. I'm planning on keeping them in Postgres syntax and calling them using EXTERNAL_QUERY.

I think the reason it's currently not working is the parameter replacement doesn't work for EXTERNAL_QUERY in BigQuery. I've only spent about 1 day on this. 

If we rewrote it to the tables in BigQuery using the context_store_entity and context_store_keymap it would incur a query cost. I'm not sure about what the cost would be.

**Comment by pushyamig** [comment]
>If we rewrote it to the tables in BigQuery using the context_store_entity and context_store_keymap it would incur a query cost. I'm not sure about what the cost would be.

From what I understood with BQ cost for running Context Store queries shouldn't be a lot. The reason as I understood from Unizin Folks is since the Storage for Context store data is not as big as BQ Event store so we don't have query Processing cost as with event store. 

>I think the reason it's currently not working is the parameter replacement doesn't work for EXTERNAL_QUERY in BigQuery. I've only spent about 1 day on this.

IMHO, I feel if we are switching to BQ backend that I feel the queries should also be BQ Fomat otherwise we are still be dealing with 2 different system(Postgres, BQ). But I fully don't understand why parameter replacement doesn't work so I am just saying what I am thinking. Since last time I did the update I was dealing with Postgres, Bigquery connector and it was not fun updating.

**Comment by jonespm** [comment]
I added some more thoughts to this on Slack. I felt this was possibly the easiest way for a first pass. If we wanted to push this forward and do a rewrite to 100% BigQuery we might decide to use the canvas (CD2) tables instead of the context_store. It just feels better to only be going to one source for everything even if BigQuery is just mostly a "proxy" for the PostGres data.

**Comment by jonespm** [comment]
Thanks for the feedback here. Because of complications around named parameters and the temporary table, I've decided to take your suggestion and explore rewriting this directly into BigQuery. Currently it runs as far as the assignment table so it's getting there!

**Comment by pushyamig** [comment]
yes, all the tables other than Submission should be simple, just syntax change and probably some the way the variables are queried.


With submission query, with BQ we have to check if we want to create a Temp table route, I felt BQ is might be good with handling stuff. But Only way to test is in Myla Beta since it has lot of courses.

**Comment by jonespm** [comment]
Doesn't seem like temp table is necessary anymore. Can just create it using with a WITH and CTE which seems to be creating a fast temporary table now. Was only necessary back in PostGres.

**Comment by pushyamig** [comment]
I will review this today or tomorrow

File: dashboard/common/db_util.py
Comment (zqian) [inline dashboard/common/db_util.py:200]: Cannot find the use of this function in the PR, though the cron runs successfully with this PR updates.

File: dashboard/cron.py
Comment (zqian) [inline dashboard/cron.py:686]: We need to watch for the cost. I have 20 MyLA courses in localhost, and the initial run with this PR cost $18, and subsequent run cost $0.1

Comment (pushyamig) [reply dashboard/cron.py:686]: I only has 4 course and costed $24 next run cost $0.1. I would expect to cost less
```
// This is the total row count
resources_access:	2115
resources:	248
submissions:	1759
assignments:	95
user:	100
course:	5
```

**Comment by pushyamig** [comment]
I feel this PR should be run from MyLA Beta and see how long it take to run. since MyLA Beta has about 50+ course

File: dashboard/common/db_util.py
Comment (pushyamig) [inline dashboard/common/db_util.py:200]: Just curious, why is this function needed? with postgres this was not needed.

**Comment by pushyamig** [comment]
The PR seems be working from fetching results as expected. I have some minor comment. But I feel this PR should be run against MyLA Beta course and see how it performance specially with Submission Query.

File: dashboard/settings.py
Comment (jonespm) [reply dashboard/settings.py:355]: Yeah, without this Pylance reports that this import is missing and it is. This is in a try/catch and it's expected this file might not exist. I'm not sure if we added this feature, but I just wanted to remove the error from the log and this seems like the best suggestion to turn it off here.

File: config/env_sample.hjson
Comment (jonespm) [reply config/env_sample.hjson:188]: It's not using an external query anymore, it's retrieved directly from tables in the UDP

File: dashboard/common/db_util.py
Comment (jonespm) [reply dashboard/common/db_util.py:186]: Yes, this should probably also just be removed as it's useless code now and will always return the default of `{}`. I also see this in the cron.py.

File: dashboard/cron.py
Comment (jonespm) [reply dashboard/cron.py:686]: The cost is high likely have courses without a start date, without a term or with a really old term in your database. It might also be a bug with retrieving the dates for terms which wouldn't be good. The event store is getting huge now so doing a full query is multiple TB of data. 

I think we should have a safety case in here so it doesn't retrieve more than X months of data from the `event_store` unless explicitly directed to. Generally there won't ever be a case to populate old course data.

**Comment by jonespm** [comment]
> I feel this PR should be run from MyLA Beta and see how long it take to run. since MyLA Beta has about 50+ course

Do you mean to use the course-set that's on the MyLA beta server? That could be done locally by just exporting the course table. It should run about the same local or on that server and uses the same data set.

File: dashboard/cron.py
Comment (jonespm) [reply dashboard/cron.py:45]: I remember that I had this in there because I was thinking of switching the way we configure BigQuery to use an explicit JSON value in the hjson rather than the bq_cred.json file. This is probably a separate issue but something I noticed in the newer version of the library that is supported.

File: dashboard/common/db_util.py
Comment (jonespm) [reply dashboard/common/db_util.py:200]: This is needed because BigQuery has special types in it's JobConfig that have to be passed explicitly to bind parameters. 

All of the bindings could have been changed and this would have been unneeded but I wasn't sure which method would have been easier. This is more compatibility for the old code. For instance I've got a call like this which isn't changed much from the old code.

```
        query_params = {
                        'course_ids': self.valid_locked_course_ids,
                        'canvas_data_id_increment': settings.CANVAS_DATA_ID_INCREMENT,
                        }
        df = self.execute_bq_query(self.queries['submission'], query_params).to_dataframe()
```

This could have been rewritten completely in BigQuery style to pass a `QueryJobConfig `object explicitly. 

```
job_config = bigquery.QueryJobConfig(
    query_parameters=[
        bigquery.ArrayQueryParameter('course_ids', 'INT64', self.valid_locked_course_ids,),
        bigquery.ScalarQueryParameter('canvas_data_id_increment', 'INT64', settings.CANVAS_DATA_ID_INCREMENT),
    ]
)

df = self.execute_bq_query(self.queries['submission'], query_params).to_dataframe()
```

Perhaps this would have just been a better idea to do this? I could still do this and remove this helper function as there's only 5 places this was done. I thought it would have been more work to change and it would have to support PostGres syntax still, but that was before this all changed fully to BigQuery. It just requires you to learn more about the special BigQuery types rather than infer them.

**Comment by pushyamig** [comment]
>Do you mean to use the course-set that's on the MyLA beta server? That could be done locally by just exporting the course table. It should run about the same local or on that server and uses the same data set.

I don't think I care if it's done locally or Beta server. If you run the courses from MyLA Beta or Test (Test has more courses) and see how the cron performance in same way as current prod then that should be good.

**Comment by jonespm** [comment]
> > Do you mean to use the course-set that's on the MyLA beta server? That could be done locally by just exporting the course table. It should run about the same local or on that server and uses the same data set.
> 
> I don't think I care if it's done locally or Beta server. If you run the courses from MyLA Beta or Test (Test has more courses) and see how the cron performance in same way as current prod then that should be good.

That makes sense, I was only testing it locally on a few courses but even that ran significantly faster now. I should be able to try the full list of courses.

File: dashboard/common/db_util.py
Comment (jonespm) [reply dashboard/common/db_util.py]: I have made this change in the latest version of the code.

**Comment by jonespm** [comment]
You should only test this with a current term course for now as it will pull down too much data otherwise on an empty database.

**Review by pushyamig — APPROVED** [review]
Nice Work @jonespm for pulling this off in short amount of time.

**Comment by jonespm** [comment]
Thanks, didn't notice the checks, they were valid about unused imports so should pass now.

**Comment by jonespm** [comment]
I'm not sure if I'll have time to run on the test environment today this so planning on waiting until Monday to merge and test this out then.

**Comment by zqian** [comment]
cron job runs successfully on test Jenkins.

**Pull request merged**
